### PR TITLE
adds the ability to "mix" solutions (reactions caused by using an item on a solution holder)

### PR DESCRIPTION
**Pull request by timothyteakettle**
adds a component for mixing, that can take a list of strings that are essentially mixing 'categories'
solutions have a canMix variable, and there's a new method for getting a mixable solution from an entity
mixing system attempts to update chemicals for entities hit by an entity that has the mixing component
reactions can have a list of strings that are the required mixing categories for the component to have

i.e. if a bible has Holy, and the blood to wine reaction has Holy, then hitting a beaker of blood with a bible will create wine (this is also a reaction added as an example!)

solutions default to not allowing mixing by default, and then i set beakers cups and bottles to allow mixing, because i feel like it's easier and more maintainable to say what does allow mixing compared to what does not

personally i think it has the best use for cooking/bartending because you could actually mix a drink using a spoon, for example, or have a shaker that when used on itself (this would probably need a second event adding to the mixer system but is a very minor change!) mixes reactions inside it

made in response to suggestion #11778

tested locally to make sure it works as intended (both using the recipe correctly, and trying to mix it with something that has the wrong mixing category in the component such that it doesn't mix, and trying to mix things that don't allow mixing)

this is my first PR for ss14 so any feedback on alternative approaches, code quality, etc would be greatly appreciated

- [ ] I have added screenshots/videos to this PR showcasing its changes ingame
- [X] This PR does not require an ingame showcase

:cl:
- add: ability to mix reactions with certain entities, included example being blood can be converted into wine using a bible at a ratio of 1:1

### CODE REVIEW

File: Content.Server/Chemistry/EntitySystems/SolutionContainerSystem.cs [hunk 1]
```diff
@@ -333,6 +335,27 @@ public FixedPoint2 GetReagentQuantity(EntityUid owner, string reagentId)
         return reagentQuantity;
     }
 
+    public bool TryGetMixableSolution(EntityUid uid,
+    [NotNullWhen(true)] out Solution? solution,
+    SolutionContainerManagerComponent? solutionsMgr = null)
```
Comment (moonheart08) [inline Content.Server/Chemistry/EntitySystems/SolutionContainerSystem.cs:340]: Wrong indentation, check how your IDE wants this indented assuming it's not visual studio. (If it is, ill look for a correct example later)

File: Content.Server/Chemistry/EntitySystems/SolutionContainerSystem.cs [hunk 2]
```diff
@@ -333,6 +335,27 @@ public FixedPoint2 GetReagentQuantity(EntityUid owner, string reagentId)
...
+    {
+        if (!Resolve(uid, ref solutionsMgr, false))
+        {
+            solution = null;
+            return false;
+        }
+
+        var tryGetSolution = solutionsMgr.Solutions.FirstOrNull(x => x.Value.CanMix);
```
Comment (moonheart08) [inline Content.Server/Chemistry/EntitySystems/SolutionContainerSystem.cs:348]: Would be worth first firing an event in the target to try and let the target find the solution for us. (even if we don't use this right now it'll 100% be used in the future)

**Review by moonheart08 — CHANGES_REQUESTED** [review]
This doesn't seem to give any end user feedback.

Also maybe best for it to fire an event on both the mixer and mixed entities before doing anything to check for cancellation, a separate event afterward so they can play sounds/etc.

File: Content.Server/Chemistry/EntitySystems/SolutionContainerSystem.cs [hunk 3]
```diff
@@ -333,6 +335,27 @@ public FixedPoint2 GetReagentQuantity(EntityUid owner, string reagentId)
...
+        if(tryGetSolution.HasValue)
```
Comment (0x6273) [inline Content.Server/Chemistry/EntitySystems/SolutionContainerSystem.cs:349]:
```suggestion
        if (tryGetSolution.HasValue)
```

File: Content.Shared/Chemistry/Reaction/ReactionMixerComponent.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+{
+    /// <summary>
+    ///     A list of IDs for categories of reactions that can be mixed (i.e. HOLY for a bible, DRINK for a spoon)
+    /// </summary>
+    [ViewVariables]
+    [DataField("reactionTypes")]
+    public List<string> ReactionTypes = default!;
+
+
+}
```
Comment (0x6273) [inline Content.Shared/Chemistry/Reaction/ReactionMixerComponent.cs:14]:
```suggestion
    public List<string> ReactionTypes = default!;
}
```

File: Content.Shared/Chemistry/Reaction/SharedChemicalReactionSystem.cs [hunk 5]
```diff
@@ -115,6 +115,15 @@ private bool CanReact(Solution solution, ReactionPrototype reaction, EntityUid o
                 return false;
             }
 
+
+
+            if((mixerComponent == null && reaction.MixingCategories != null) ||
```
Comment (0x6273) [inline Content.Shared/Chemistry/Reaction/SharedChemicalReactionSystem.cs:120]:
```suggestion
            }

            if((mixerComponent == null && reaction.MixingCategories != null) ||
```

**Comment by timothyteakettle** [comment]
thanks for the feedback, i'll work on addressing those points now!

File: Content.Server/Chemistry/EntitySystems/SolutionContainerSystem.cs (hunk 1 above)
Comment (timothyteakettle) [reply Content.Server/Chemistry/EntitySystems/SolutionContainerSystem.cs:340]: fixed!

File: Content.Server/Chemistry/EntitySystems/SolutionContainerSystem.cs (hunk 2 above)
Comment (timothyteakettle) [reply Content.Server/Chemistry/EntitySystems/SolutionContainerSystem.cs:357]: did this via GetMixableSolutionAttemptEvent

**Review by moonheart08 — APPROVED** [review]

**Pull request merged**